### ISSUE TITLE
EXPORT_ES6 doesn't work with USE_PTHREADS

### ISSUE THREAD

**Issue by VirtualTim** [emscripten-core/emscripten#8176]
In shell_pthreads.js we have the line `ENVIRONMENT_IS_PTHREAD = false` ([here](https://github.com/emscripten-core/emscripten/blob/incoming/src/shell_pthreads.js#L10)). ES6 modules use strict mode, which [disallows assigning a value to an undeclared variable](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Undeclared_var).

As someone coming from native code to JS I'm not really sure how to fix this. Hopefully it's simple?

**Comment by VirtualTim** [comment]
Related to #1659.

**Comment by kripken** [comment]
It's declared in another file, src/worker.js - but yeah, I think that just exists for workers, and so it may be undeclared on the main thread. Adding a `var` for it in src/preamble.js, in the ifdefed code for pthreads, is probably the right thing.

**Comment by VirtualTim** [comment]
So I added a `var ENVIRONMENT_IS_PTHREAD;` under the `USE_PTHREADS` [here](https://github.com/emscripten-core/emscripten/blob/incoming/src/preamble.js#L35). Unfortunately that didn't fix it and the code hangs joining the threads. I'm guessing that `emscripten_main_thread_process_queued_calls` tries to execute on the wrong thread or something.

Example Code I'm using:
```
#include <stdio.h>
#include <thread>
#include <emscripten.h>

class test_wrapper {
private:
	static void print_MutexLock(int id) {
		printf("thread number is %d\n", id);
	}

public:
	void EMSCRIPTEN_KEEPALIVE pthread_test(int numThreads) {
		std::thread threadPool[numThreads];
		for (int i=0; i<numThreads; i++)
				threadPool[i] = std::thread(print_MutexLock, i);
	 
		for (auto& thread : threadPool)
			thread.join();	//hang here
	}
};

#include "sdk_idl_glue.cpp"
```
test_wrapper.js:
```
import { default as mySDK} from './test.js';
console.log('loaded wrapper');
var SDKinstance = mySDK().then(function(SDKinstance) {
	var test_wrapper_instance = new SDKinstance.test_wrapper();
	test_wrapper_instance.pthread_test(4);
});
```
test.html:
replace the inline script with `<script type='module' src="./test_wrapper.js"></script>`

**Comment by sbc100** [comment]
Creating and then synchronously joining threads from the main thread can be problematic. 

Firstly the creation part.  Unless you set PTHREAD_POOL_SIZE then then new worker threads are needed which (I think) won't actually spin up until you yield back to the event loop.   Secondly joining is a blocking operation which is dangerous on the main thread because you won't be able to run the event loop while you wait which can cause a deadlock if the thread you are join is waiting on main thread work.

So, for this example, I would try removing the thread.join() calls completely and setting PTHREAD_POOL_SIZE to 4.

In other words, I think your failure might not related to EXPORT_ES6.  Can you confirm this code works when not uses as a module?

Finally I don't understand how the test_wrapper constructor is exported to JS in your example... it seems like you would need embind or similar to make this work.

**Comment by VirtualTim** [comment]
Oh yeah, the constructor was exported using webidl_binder. I didn't include the idl code because I didn't want to overcomplicate things.
This code was just copy-pasted from a previously working example, but I guess things could have changed in Emscripten since then. Anyway, I replaced the `join` with a `detach`, and it doesn't hang, but throws a bunch of errors. Seems like the issue is in worker.js. `e.data.urlOrBlob === 'undefined'`, so the following code fails/throws:
```
if (typeof e.data.urlOrBlob === 'string') {
        importScripts(e.data.urlOrBlob);
      } else {
        var objectUrl = URL.createObjectURL(e.data.urlOrBlob);
}
```
I'll do some more experimenting today.

**Comment by VirtualTim** [comment]
I've added a WIP RP, because seeing the changes are probably easier than reading paragraphs to figure out what I'm talking about. #8306.
